openroad_interface/estimation.py

```diff
@@ -0,0 +1,120 @@
+import re
+import math
+import os
+
+import numpy as np
+import networkx as nx
+
+from var import directory
+
+def total_euclidean_distance(net_list: list, coord_data: dict, unit: float):
+    '''
+    takes in the net list and finds the corresponding coordinate and calculates the euclidean distance from the 
+    first component of the net list. it also stores these into a list edge_list for resistance calculation purposes. 
+    '''
+    the_component = net_list[0]
+    result = 0.0
+    edge_list = []
+    for x in range(1, len(net_list)):
+        edge = math.sqrt(pow(coord_data[the_component]["x"]/unit - coord_data[net_list[x]]["x"]/unit, 2) + pow(coord_data[the_component]["y"]/unit - coord_data[net_list[x]]["y"]/unit, 2))
+        edge_list.append(edge)
+        result += edge
+    return result, edge_list
+
+def global_estimation():
+    wire_global_file = directory + "results/wire_length_global.txt"
+    global_pattern = r"grt:\s_[0-9]+_\s[0-9]*\.[0-9]+\s[0-9]+"
+
+    global_length_data = []
+    wire_global_data = open(wire_global_file)
+    wire_global_lines = wire_global_data.readlines()
+    for line in wire_global_lines:
+        if re.search(global_pattern, line) != None:
+            match = re.search(global_pattern, line)
+            globa_length = match.group(0).split()[2]
+            global_length_data.append(float(globa_length))
+    return global_length_data
+
+def parasitic_estimation(graph, design_name: str, net_out_dict: dict, node_output: dict, lef_data: dict) -> dict:
+    '''
+    calculates the resistance and capacitance using the euclidean distance and generates a .gml file with the edge attributes
+
+    param:
+        graph: modified networkx graph 
+        design_name: design name for naming the file 
+        net_out_dict: dict that lists nodes and thier respective edges (all nodes have one output)
+        node_output: dict that lists nodes and their respective output nodes
+        lef_data: dict with lef file information (res, cap, width, units)
+
+    return:
+        dict: all res cap length data
+    '''
+    final_def_file = directory + "results/final_generated-tcl.def"
+    pattern = r"_\w+_\s+\w+\s+\+\s+PLACED\s+\(\s*\d+\s+\d+\s*\)\s+\w+\s*;"
+    net_pattern =  r'-\s(_\d+_)\s((?:\(\s_\d+_\s\w+\s\)\s*)+).*'
+    component_pattern = r'(_\w+_)'
+
+    units = lef_data["units"]
```

Comment:
The only information we're using from the def file here is the placement location right?
For the estimation we shouldn't be using anything else from openroad